File: 99-recover-binary-search-tree/99-recover-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode *first = NULL, *second = NULL, *prev = NULL;
    void recoverTree(TreeNode* root) {
        swapval(root);
        int temp = first->val;
        first->val = second->val;
        second->val = temp;
    }
    
    void swapval(TreeNode* curr) {
        if (!curr)
            return;
        swapval(curr->left);
        if (prev && prev->val > curr->val) {
            if (!first)
                first = prev;
            second = curr;
        }
        prev = curr;
        swapval(curr->right);
    }
};

File: 99-recover-binary-search-tree/99-recover-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode x = null, y = null, z = null;
    public void recoverTree(TreeNode root) {
        helper(root);
        swap(x, y);
    }
    
    private void helper(TreeNode node) {
        if (node == null)
            return;
        helper(node.left);
        if (z != null && node.val < z.val) {
            y = node;
            if (x == null)
                x = z;
            else
                return;
        }
        z = node;
        helper(node.right);
    }
    
    private void swap(TreeNode x, TreeNode y) {
        int temp = x.val;
        x.val = y.val;
        y.val = temp;
    }
}